lib/SILOptimizer/Transforms/AllocBoxToStack.cpp

```diff
@@ -796,95 +900,132 @@ specializePartialApply(SILOptFunctionBuilder &FuncBuilder,
+  auto ApplyInst = Apply.getInstruction();
+  SILBuilderWithScope Builder(ApplyInst);
+
+  // Build the function_ref and ApplySite.
+  SILValue FunctionRef = Builder.createFunctionRef(Apply.getLoc(), ClonedFn);
+  switch (Apply.getKind()) {
+  case ApplySiteKind::PartialApplyInst: {
+    auto *PAI = cast<PartialApplyInst>(ApplyInst);
+    return Builder.createPartialApply(
+        Apply.getLoc(), FunctionRef, Apply.getSubstitutionMap(), Args,
+        PAI->getType().getAs<SILFunctionType>()->getCalleeConvention(),
+        PAI->isOnStack(),
+        GenericSpecializationInformation::create(ApplyInst, Builder));
+  }
+  case ApplySiteKind::ApplyInst:
+    return Builder.createApply(
+        Apply.getLoc(), FunctionRef, Apply.getSubstitutionMap(), Args,
+        GenericSpecializationInformation::create(ApplyInst, Builder));
+  case ApplySiteKind::BeginApplyInst:
+    return Builder.createBeginApply(
+        Apply.getLoc(), FunctionRef, Apply.getSubstitutionMap(), Args,
+        GenericSpecializationInformation::create(ApplyInst, Builder));
+  case ApplySiteKind::TryApplyInst: {
+    auto TAI = cast<TryApplyInst>(Apply);
+    return Builder.createTryApply(
+        Apply.getLoc(), FunctionRef, Apply.getSubstitutionMap(), Args,
+        TAI->getNormalBB(), TAI->getErrorBB(),
+        GenericSpecializationInformation::create(ApplyInst, Builder));
+  }
+  }
 }
 
-static void rewritePartialApplies(AllocBoxToStackState &pass) {
-  llvm::DenseMap<PartialApplyInst *, ArgIndexList> IndexMap;
+static void rewriteApplySites(AllocBoxToStackState &pass) {
+  llvm::DenseMap<ApplySite, ArgIndexList> IndexMap;
+  llvm::SmallVector<ApplySite, 8> AppliesToSpecialize;
   ArgIndexList Indices;
 
-  // Build a map from partial_apply to the indices of the operands
+  // Build a map from the ApplySite to the indices of the operands
   // that will be promoted in our rewritten version.
   for (auto *O : pass.PromotedOperands) {
-    auto CalleeArgIndexNumber = ApplySite(O->getUser()).getCalleeArgIndex(*O);
+    auto User = O->getUser();
+    auto Apply = ApplySite(User);
+
+    auto CalleeArgIndexNumber = Apply.getCalleeArgIndex(*O);
 
     Indices.clear();
     Indices.push_back(CalleeArgIndexNumber);
 
-    auto *PartialApply = cast<PartialApplyInst>(O->getUser());
-    llvm::DenseMap<PartialApplyInst *, ArgIndexList>::iterator It;
-    bool Inserted;
-    std::tie(It, Inserted) = IndexMap.insert(std::make_pair(PartialApply,
-                                                            Indices));
-    if (!Inserted)
-      It->second.push_back(CalleeArgIndexNumber);
+    auto iterAndSuccess = IndexMap.try_emplace(Apply, Indices);
+    // AllocBoxStack opt promotes boxes passed to a chain of applies when it is
+    // safe to do so. All such applies have to be specialized to take pointer
+    // arguments instead of box arguments This has to be done in dfs order.
+    // PromotedOperands is already in dfs order. Build AppliesToSpecialize in
```

Comment:
I would make it clear that is it already in dfs order due to how it is constructed by find**. Someone reading this code needs a back reference to the code above.